src/import/trace-event.ts

```diff
@@ -365,50 +454,166 @@ function eventListToProfileGroup(events: TraceEvent[]): ProfileGroup {
-                  eEventQueue[i] = temp
-                  swapped = true
-                  break
-                }
-              }
-            }
-
-            // If swapped is still false at this point, it means we're about to
-            // pop a stack frame that doesn't even match by name. Bummer.
           }
 
-          const e = eEventQueue.shift()!
-
-          tryToLeaveFrame(e)
-          break
+          // If swapped is still false at this point, it means we're about to
+          // pop a stack frame that doesn't even match by name. Bummer.
         }
 
-        default:
-          const _exhaustiveCheck: never = queueName
-          return _exhaustiveCheck
+        const e = eEventQueue.shift()!
+
+        tryToLeaveFrame(e)
+        break
       }
+
+      default:
+        const _exhaustiveCheck: never = queueName
+        return _exhaustiveCheck
     }
+  }
+
+  for (let i = frameStack.length - 1; i >= 0; i--) {
+    const frame = frameInfoForEvent(frameStack[i])
+    console.warn(`Frame "${frame.key}" was still open at end of profile. Closing automatically.`)
+    profileBuilder.leaveFrame(frame, profileBuilder.getTotalWeight())
+  }
+
+  return profileBuilder.build()
+}
+
+/**
+ * Returns an array containing the time difference in microseconds between the current
+ * sample and the next sample
+ */
+function getTimeDeltasForSamples(samples: Sample[]) {
+  const timeDeltas: number[] = []
+  let lastTimeStamp = Number(samples[0].ts)
+
+  samples.forEach((sample: Sample, idx: number) => {
+    if (idx === 0) return
+
+    const timeDiff = Number(sample.ts) - lastTimeStamp
+    lastTimeStamp = Number(sample.ts)
+    timeDeltas.push(timeDiff)
+  })
 
-    for (let i = frameStack.length - 1; i >= 0; i--) {
-      const frame = frameInfoForEvent(frameStack[i])
-      console.warn(`Frame "${frame.key}" was still open at end of profile. Closing automatically.`)
-      profile.leaveFrame(frame, profile.getTotalWeight())
+  timeDeltas.push(0)
+
+  return timeDeltas
+}
+
+/**
+ * The chrome json trace event spec only specifies name and category
+ * as required stack frame properties
+ *
+ * https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.b4y98p32171
+ */
+function frameInfoForSampleFrame({name, category}: StackFrame): FrameInfo {
+  return {
+    key: `${name}:${category}`,
+    name: name,
+  }
+}
+
+function getActiveFramesForSample(
+  stackFrames: TraceWithSamples['stackFrames'],
+  frameId: number,
+): FrameInfo[] {
+  const frames = []
+  let parent: number | undefined = frameId
+
+  while (parent) {
+    const frame: StackFrame = stackFrames[parent]
+
+    if (!frame) {
+      throw new Error(`Could not find frame for id ${parent}`)
+    }
+
+    frames.push(frameInfoForSampleFrame(frame))
+    parent = frame.parent
+  }
+
+  return frames.reverse()
+}
+
+function constructProfileFromSampleList(
+  contents: TraceWithSamples,
+  samples: Sample[],
+  name: string,
+) {
```

Comment:
Add a return type to the function signature please

I have the eslint rule dispabled to always enforce return types being required because sometimes it's absurdly obvious or annoying to do, but here not's not trivially obvious.

Also for consistency, please rename to `sampleListToProfile`

Reply:
Will add a type - there are four functions:

- `eventListToProfileGroup` which calls `constructProfileFromTraceEvents` for each thread
- `sampleListToProfileGroup` which calls `constructProfileFromSampleList` for each thread

Would you like me to rename to `eventListToProfile` and `sampleListToProfile` - IMO them being so similar in name makes them easy confuse.

Reply:
Updated [here](c8c1e7f1af36f5fac5d62ee268ac62c56e3eda8b), seems sensible